keps/sig-api-machinery/5589-gogo-dependency/README.md

```diff
@@ -0,0 +1,640 @@
+
+* v1.2: original protobuf support proposal [PR](https://github.com/kubernetes/kubernetes/pull/22600)
+([final archive location](https://github.com/kubernetes/design-proposals-archive/blob/main/api-machinery/protobuf.md)).
+* v1.2: generate .proto files from Go types ([#17854](https://github.com/kubernetes/kubernetes/pull/17854)),
+        write back proto tags to Go types ([#19426](https://github.com/kubernetes/kubernetes/pull/19426))
+* v1.3: add protobuf serializer ([#20377](https://github.com/kubernetes/kubernetes/pull/20377))
+* v1.4: customize String() ([#28815](https://github.com/kubernetes/kubernetes/pull/28815))
+* v1.7: make map serialization deterministic ([#47701](https://github.com/kubernetes/kubernetes/pull/47701))
+* 2020: gogo / golang/protobuf incompatibilities stall dependency updates, panic etcd libraries, get resolved in a best-effort way
+* 2021-2022: gogo transitions to deprecated / unmaintained
+* v1.35: This KEP proposed
+
+## Drawbacks
+
+Taking on responsibility for complex protobuf generation code is not ideal.
+However, I would claim that the Kubernetes project already has implicit responsibility
+for the *output* of the unmaintained generator, so I don't think a maintenance-mode fork
+whose only consumer is the Kubernetes API protobuf generator is significantly different
+than the current state.
+
+## Alternatives
+
+Two alternatives were considered:
+
+1. Switching to standard `protoc`/`protoc-gen-go` generation.
+   The impact to non-protobuf use of Kubernetes API go types,
+   and the mismatch between Kubernetes requirements around canonical serialization
+   and dropping of unknown fields ruled this out. See the [Notes/Constraints/Caveats](#notesconstraintscaveats)
+   section for more details.
+
+2. Switching to another non-standard protobuf generation (e.g. `vtprotobuf`).
+   This would have required significantly more work to achieve,
+   and would have been significantly more difficult to verify compatibility.
+   There were not clear benefits that would justify this amount of time investment.
+   The approach described in this KEP does not preclude such a change in the future,
+   if it becomes apparent there are benefits that would justify a change of that magnitude,
+   and a plan is proposed to make that change safely.
+
+## Possible Future Work
+
+The following are *possible* future changes that could be made to the protobuf generator
+once it is in a location that can be modified in ways scoped to Kubernetes API types.
+These changes are not planned as part of this KEP, but are captured here for reference.
+
+**Add omitzero support**
+
+Current protobuf serialization *always* outputs non-pointer scalar and struct fields.
+This means that feature-gate-disabled fields must be created as pointers so they will not be serialized unless populated.
+Adding support for an `omitzero` tag to protobuf serialization would allow for non-pointer fields which can still be feature-gate-disabled.
```

Comment:
+1

I proposed that as an option at one point on Slack when we discussed the need to use pointers for new fields.